src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,317 @@
+        {
+            if (mEntries[i].Compare(value))
+            {
+                return Remove(i);
+            }
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes all the entries in the table.
+     * @return CHIP_NO_ERROR on success.
+     */
+    CHIP_ERROR RemoveAll()
+    {
+        ReturnErrorOnFailure(this->Load());
+        for (uint16_t i = 0; i < mLimit; ++i)
+        {
+            OnEntryRemoved(mEntries[i]);
+        }
+        mCount = 0;
+        ReturnErrorOnFailure(this->Save());
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Check if the table is empty
+     * @return True when there is no entry in the table. False if there is at least one
+     */
+    bool IsEmpty()
+    {
+        VerifyOrReturnError(CHIP_NO_ERROR == this->Load(true), true);
+        return 0 == this->mCount;
+    }
+
+protected:
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        // Count
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(PersistentTags::kCount), this->mCount));
+
+        // Entries
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(PersistentTags::kList), TLV::kTLVType_Array, list));
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
+        {
+            ReturnErrorOnFailure(mEntries[i].Serialize(writer));
+        }
+        ReturnErrorOnFailure(writer.EndContainer(list));
+
+        CHIP_ERROR err = writer.EndContainer(container);
+        return err;
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        // Count
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kCount)));
+        ReturnErrorOnFailure(reader.Get(this->mCount));
+
+        // Entries
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::ContextTag(PersistentTags::kList)));
+        ReturnErrorOnFailure(reader.EnterContainer(list));
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
+        {
+            ReturnErrorOnFailure(mEntries[i].Deserialize(reader));
+        }
+        ReturnErrorOnFailure(reader.ExitContainer(list));
+
+        return reader.ExitContainer(container);
+    }
+
+    virtual void OnEntryAdded(const EntryType & entry) {}
+    virtual void OnEntryRemoved(const EntryType & entry) {}
+    virtual void OnEntryModified(const EntryType & old_value, const EntryType & new_value) {}
+
+private:
+    // EntryType & At(size_t index) { return (index < this->mLimit) ? mEntries[index] : mEntries[0]; }
+    EntryType mEntries[kMaxListSize];
+    const size_t mLimit = 0;
```

Comment:
We end up setting mLimit to kMaxListSize in the constructor.... but why not do it here?  kMaxListSize is a template parameter, so a compile-time known constant....

More to the point, why do we need to store it in the member at all if it's a compile constant attached to the class?